### PR TITLE
[REVIEW]Add support for grouping by Series of arbitrary length

### LINKED ISSUE #2159: [FEA] Allow grouping by Series of arbitrary length
A specific case of #1477, we need to allow grouping an object by a Series of different length from the object. For example, Pandas allows:

```
In [5]: a = pd.DataFrame({'a': [1, 1, 2], 'b': [2, 3, 4]}, index=[4, 5, 6])                             

In [6]: b = pd.Series([1, 2, 2], index=[3, 4, 5])   

In [7]: a.groupby(b).sum()                          
Out[7]: 
     a  b
2.0  2  5
```

### LINKED ISSUE #2874: [BUG]groupby issue when series in `by` has same name as one of the columns in dataframe
**Describe the bug**
cudf.groupby returns wrong result if the name of one of the columns in DataFrame is same as the name of the column Series passed in `by`.

**Steps/Code to reproduce bug**
```
import cuds
a = cudf.DataFrame({'a': [1, 1, 2], 'b': [2, 3, 4]}, index=[4, 5, 6])
b = cudf.Series([1, 2, 2], name='a', index=[4, 5, 6])   
a.groupby(b).sum()
```
```
a	b
	
1	2
2	7

```
**Expected behavior**
```
import pandas as pd
a = pd.DataFrame({'a': [1, 1, 2], 'b': [2, 3, 4]}, index=[4, 5, 6])
b = pd.Series([1, 2, 2], name='a', index=[3, 4, 5])   
a.groupby(b).sum()
```

```
a     b

2     5
```

**Environment overview (please complete the following information)**
 
 - Method of cuDF install: From source [0.10]

### PR DESCRIPTION
**Pull request by rgsl888prabhu**
The code change is to get inner join of the `Series` and `DataFrame` and use groupie on it.

close #2159
close #2874

### CODE REVIEW

**Comment by rgsl888prabhu** [comment]
rerun tests

**Comment by rgsl888prabhu** [comment]
@shwina @kkraus14

File: python/cudf/cudf/tests/test_groupby.py [hunk 1]
```diff
@@ -445,20 +445,6 @@ def test_groupby_external_series(series):
     assert_eq(pxx, gxx)
 
 
-@pytest.mark.xfail(
-    raises=NotImplementedError,
-    reason="cuDF does not support arbitrary series "
-    "index lengths for groupby",
-)
-@pytest.mark.parametrize("series", [[0, 1], [1, 1, 1, 1]])
```
Comment (shwina) [inline python/cudf/cudf/tests/test_groupby.py]: Did this test not pass? Just curious why we're deleting it

Comment (rgsl888prabhu) [reply python/cudf/cudf/tests/test_groupby.py]: It was expecting a exception, but now we will not come across this issue as we support arbitrary length.

Comment (shwina) [reply python/cudf/cudf/tests/test_groupby.py]: Now that it's implemented, you should just remove the `xfail` and keep this test.

File: python/cudf/cudf/core/dataframe.py [hunk 2]
```diff
@@ -2539,6 +2539,19 @@ def groupby(
             raise TypeError(
                 "groupby() requires either by or level to be" "specified."
             )
+        df = self
+        if isinstance(by, Series):
```
Comment (shwina) [inline python/cudf/cudf/core/dataframe.py]: We should be able to group by arbitrary sequences (not just `Series` objects), so I'm not sure if this is the best place to do this. Inside the groupby code somewhere we convert `by` to a `Series` and that's where this should happen.

Comment (shwina) [reply python/cudf/cudf/core/dataframe.py]: I take this back. It looks like Pandas only allows grouping by `Series` if the length of the grouper is different from the dataframe

Comment (rgsl888prabhu) [reply python/cudf/cudf/core/dataframe.py]: And if it is a sequence, it normally expects it to be columns names

Comment (shwina) [reply python/cudf/cudf/core/dataframe.py]: Hmm - this still won't do. In general, you can group by multiple `Series` objects, or a combination of columns and `Series` objects:

```
In [64]: a = pd.DataFrame({'a': [1, 1, 1, 2, 2], 'b': [1, 2, 3, 4, 5]})                                                                                                                                         

In [65]: a.groupby(['a', pd.Series([1, 2, 3])])                                                                                                                                                                 
Out[65]: <pandas.core.groupby.generic.DataFrameGroupBy object at 0x7f9097a0a400>
```

Comment (rgsl888prabhu) [reply python/cudf/cudf/core/dataframe.py]: I didn't realize this scenario, then I have handle it as you had mentioned @ https://github.com/rapidsai/cudf/blob/b96073c7b481087f009849eee527cfe78ac8c081/python/cudf/cudf/core/groupby/groupby.py#L201

**Comment by rgsl888prabhu** [comment]
@shwina

File: python/cudf/cudf/core/groupby/groupby.py [hunk 3]
```diff
@@ -528,3 +534,51 @@ def _add_prefixes(names, prefixes):
         for i, (prefix, col_name) in enumerate(zip(prefixes, names)):
             prefixed_names[i] = f"{prefix}_{col_name}"
     return prefixed_names
+
+
+def _align_by_and_df(obj, by, how="inner"):
+    """
+    Returns a pair of dataframes and a list may be containing
```
Comment (shwina) [inline python/cudf/cudf/core/groupby/groupby.py]: Perhaps we should include a couple of examples that make the purpose of this helper function clear?

File: python/cudf/cudf/core/groupby/groupby.py [hunk 4]
```diff
@@ -528,3 +534,51 @@ def _add_prefixes(names, prefixes):
...
+                    sr.to_frame(series_count), how="inner", sort="True"
+                )
+            series_count += 1
+
+    series_count = 0
+    if new_obj is not None:
+        new_obj = new_obj.join(obj, how=how, sort="True")
+        columns = new_obj.columns
+        for i in by:
+            if isinstance(i, cudf.Series):
```
Comment (shwina) [inline python/cudf/cudf/core/groupby/groupby.py]: As a matter of style, `i` is typically used for integer indices, how about `by_col` instead?

**Comment by shwina** [comment]
Ready to merge pending CI/style checks

**Comment by rgsl888prabhu** [comment]
@kkraus14

**Comment by kkraus14** [comment]
rerun tests

**Comment by rgsl888prabhu** [comment]
rerun tests

**Comment by kkraus14** [comment]
rerun tests

**Comment by kkraus14** [comment]
@rgsl888prabhu there's failures in dask-cudf tests

**Comment by rgsl888prabhu** [comment]
> @rgsl888prabhu there's failures in dask-cudf tests

I was looking at them, will fix it tomorrow.

**Comment by rgsl888prabhu** [comment]
The key_columns being used were just ` _column` which lost the information of the indices. So, in dask when we make partitions the indices of by would start from `0` but dataframe would retain their original `indices`, so the result was wrong.

**Comment by rgsl888prabhu** [comment]
rerun tests

**Comment by rgsl888prabhu** [comment]
rerun tests

**Comment by rgsl888prabhu** [comment]
rerun tests

**Comment by kkraus14** [comment]
rerun tests

**Comment by kkraus14** [comment]
rerun tests

**Comment by rgsl888prabhu** [comment]
rerun tests

**Pull request merged**